packages/core-ethereum/crates/core-ethereum-misc/src/commitment.rs

```diff
@@ -0,0 +1,209 @@
+}
+
+impl ChannelCommitmentInfo {
+    /// Generate the initial commitment seed using this channel information and the given
+    ///node private key.
+    pub fn create_initial_commitment_seed(&self, private_key: &[u8]) -> Result<Box<[u8]>> {
+        let mut buf = Vec::with_capacity(U256::SIZE + 4 + Hash::SIZE + self.contract_address.len());
+        buf.extend_from_slice(&self.channel_epoch.to_bytes());
+        buf.extend_from_slice(&self.chain_id.to_be_bytes());
+        buf.extend_from_slice(&self.channel_id.to_bytes());
+        buf.extend_from_slice(self.contract_address.as_bytes());
+
+        derive_commitment_seed(private_key, &buf).map_err(|e| CryptoError(e))
+    }
+}
+
+/// Initializes commitment for the given channel.
+/// The ChainCommitter is used to tell the current state of the channel and to determine if re-initialization is
+/// needed or not.
+pub async fn initialize_commitment<T, C>(
+    db: &mut T,
+    private_key: &[u8],
+    channel_info: &ChannelCommitmentInfo,
+    committer: &mut C,
+) -> Result<()>
+where
+    T: HoprCoreEthereumDbActions,
+    C: ChainCommitter,
+{
+    let contains_already = db.get_commitment(&channel_info.channel_id, 0).await?.is_some();
+    let chain_commitment = committer.get_commitment().await;
+    if contains_already && chain_commitment.is_some() {
+        match find_commitment_preimage(db, &channel_info.channel_id).await {
+            Ok(_) => return Ok(()),
+            Err(e) => {
+                warn!("Secret is found but failed to find preimage, reinitializing.. {e}")
+            }
+        }
+    }
+
+    debug!(
+        "reinitializing (db: {contains_already}, chain: {})",
+        chain_commitment.map(|h| h.to_hex()).unwrap_or("false".to_string())
+    );
+
+    create_commitment_chain(
+        db,
+        &channel_info.channel_id,
+        &channel_info.create_initial_commitment_seed(private_key)?,
+        committer,
+    )
+    .await
+}
+
+#[cfg(all(not(target_arch = "wasm32"), test))]
+mod tests {
+    use crate::commitment::{
+        bump_commitment, find_commitment_preimage, initialize_commitment, ChannelCommitmentInfo, MockChainCommitter,
+    };
+    use async_std;
+    use core_crypto::types::{Hash, PublicKey};
+    use core_ethereum_db::db::CoreEthereumDb;
+    use hex_literal::hex;
+    use utils_db::db::DB;
+    use utils_db::leveldb::rusty::RustyLevelDbShim;
+    use utils_types::primitives::U256;
+    use utils_types::traits::BinarySerializable;
+
+    const PRIV_KEY: [u8; 32] = hex!("492057cf93e99b31d2a85bc5e98a9c3aa0021feec52c227cc8170e8f7d047775");
+
+    fn create_mock_db() -> CoreEthereumDb<RustyLevelDbShim> {
+        let opt = rusty_leveldb::in_memory();
+        let db = rusty_leveldb::DB::open("test", opt).unwrap();
+
+        CoreEthereumDb::new(
+            DB::new(RustyLevelDbShim::new(db)),
+            PublicKey::from_privkey(&PRIV_KEY).unwrap(),
+        )
+    }
+
+    #[async_std::test]
+    async fn test_should_publish_hash_secret() {
```

Comment:
This could be split into 2 tests, could it not?

Reply:
I don't think that it would make sense to split it, because the second test would essentially do exactly the same flow again only to test the "repeated initializations" outcome.